### PR TITLE
Sphinx - Vlada Rapaport & Brianna Root

### PR DESCRIPTION
**Pull request by vladarap88**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -140,31 +140,31 @@ def test_adds_movie_to_non_empty_user_watchlist():
...
-@pytest.mark.skip()
+        # Arrange
+        janes_data = {
+            "watchlist": [{
+                "title": MOVIE_TITLE_1,
+                "genre": GENRE_1,
+                "rating": RATING_1
+            }],
+            "watched": []
+        }
```
Comment (yangashley) [inline tests/test_wave_01.py:153]: Indentation was changed here and on lines 156 - 160. The code should be indented 4 spaces in, but is indented 8 spaces. Be mindful of what changes get added to your commit since we don't want to introduce inconsistencies in a project.

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -140,31 +140,31 @@ def test_adds_movie_to_non_empty_user_watchlist():
...
+
+
+        # Act
+        updated_data = watch_movie(janes_data, MOVIE_TITLE_1)
+
+        # Assert
+        assert len(updated_data["watchlist"]) == 0
+        assert len(updated_data["watched"]) == 1
+        assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1 # added below asserts
+        assert updated_data["watched"][0]["genre"] == GENRE_1
```
Comment (yangashley) [inline tests/test_wave_01.py:163]: If the order of these movies were somehow changed, then the zero-th element might not be equal to `MOVIE_TITLE_1`. Do we need to specifically test that the first element of the watched list is equal to that title?

A more general way to write this assertion could look like: 
```python
assert HORROR_1 in updated_data["watched"]
```

Line 163-164 don't really affect the test, but they also aren't testing what we're interested in testing. Per the title of this test, we're looking to test that a movie moves from the watchlist to the watched list. Consider removing these two lines to make the test more concise.

File: tests/test_wave_01.py [hunk 3]
```diff
@@ -140,31 +140,31 @@ def test_adds_movie_to_non_empty_user_watchlist():
...
+        assert updated_data["watched"][0]["rating"] == RATING_1
+        assert updated_data["watchlist"] == []
```
Comment (yangashley) [inline tests/test_wave_01.py:165]: Note that we don't need to write an assertion to test that watchlist is empty because the movie was moved to the watched list because the assertion was already provided on line 160. Line 165 could be removed to keep the test concise.

Also, since you wrote an assertion to check that `HORROR_1` was in watched, you could test the opposite to make that it is __not__ in watchlist, like:

```python
assert HORROR_1 not in updated_data["watchlist"]

File: tests/test_wave_01.py [hunk 4]
```diff
@@ -182,13 +182,12 @@ def test_moves_movie_from_watchlist_to_watched():
     # Assert
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
+    assert updated_data["watched"][1]["title"] == movie_to_watch["title"] # added asserts
+    assert updated_data["watched"][1]["genre"] == movie_to_watch["genre"]
+    assert updated_data["watched"][1]["rating"] == movie_to_watch["rating"]
```
Comment (yangashley) [inline tests/test_wave_01.py:187]: Similar to my comment above, we might consider removing lines 186-187 since this test is not concerned with genre/ratings.

File: tests/test_wave_01.py [hunk 5]
```diff
@@ -182,13 +182,12 @@ def test_moves_movie_from_watchlist_to_watched():
...
+    assert updated_data["watchlist"][0] == FANTASY_1
```
Comment (yangashley) [inline tests/test_wave_01.py:188]: You check that FANTASY_1 is still in the watchlist and that HORROR_1 was added to the watched list. You might also consider checking that HORROR_1 was not added to watchlist with `assert HORROR_1 not in updated_data["watchlist"]`. 

You could also check that FANTASY_2 which was already in the watched list wasn't removed.

File: tests/test_wave_03.py [hunk 6]
```diff
@@ -54,13 +54,11 @@ def test_friends_unique_movies_not_duplicated():
 
     # Assert
     assert len(friends_unique_movies) == 3
+    assert FANTASY_4 in friends_unique_movies
+    assert HORROR_1 in friends_unique_movies
+    assert INTRIGUE_3 in friends_unique_movies
```
Comment (yangashley) [inline tests/test_wave_03.py:59]: 👍 Nice job checking that the correct movies are in `friends_unique_watched`. Consider adding an assertion to check that the movie that was added to the first friend's watched list on line 50 was not added to Amanda's watched list:

```python
assert INTRIGUE_3 not in amandas_data["watched"]

File: tests/test_wave_05.py [hunk 7]
```diff
@@ -53,12 +53,12 @@ def test_new_genre_rec_from_empty_friends():
...
     }
 
-    raise Exception("Test needs to be completed.")
-    # *********************************************************************
-    # ****** Complete the Act and Assert Portions of these tests **********
-    # *********************************************************************
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    assert len(recommendations) == 0 # added asserts
+    assert recommendations == []
```
Comment (yangashley) [inline tests/test_wave_05.py:59]: 👍 nice job writing the act portion of the test. The assertions on line 58 and 59 are essentially testing the same thing. One checks that the length is zero and the other checks that the list is empty. I'd remove line 59 to keep this test concise.

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,142 @@
...
 def create_movie(title, genre, rating):
-    pass
+    new_movie = {}
+    if not title or not genre or not rating:
+        return None
+    else:
+        new_movie ["title"] = title
+        new_movie ["genre"] = genre
+        new_movie ["rating"] = rating
+    return new_movie
```
Comment (yangashley) [inline viewing_party/party.py:12]: Great job with this method! 

A note about guard clauses - The Guard Clause pattern helps in these cases by “inverting” that way of thinking and rapidly returning false conditions and keeping the important code blocks (usually the longest) in the lowest indentation level possible. Read more [here](https://medium.com/lemon-code/guard-clauses-3bc0cd96a2d3)

Therefore, we should pull the main logic of this function out of the else block to emphasize that this is the main logic.

You can also create your dictionary with literal syntax and return it without first creating an empty dict on line 5:
```python
def create_movie(title, genre, rating):
    if not title or not genre or not rating:
        return None

    return {
        "title": title,
        "genre": genre,
        "rating": rating        
    }
```

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,142 @@
...
+
+def add_to_watched(user_data, movie):
+    user_data["watched"] += [movie]
+    return user_data
```
Comment (yangashley) [inline viewing_party/party.py:16]: We should use the `append` method over list concatenation when possible. 

Append is more readable and efficient in terms of time complexity. Using `append` clearly conveys the intention of adding an element to a list and is faster because it modifies a list in place instead of creating a new list (which `+=` does)

You can read more about this in this [SO thread](https://stackoverflow.com/questions/2022031/python-append-vs-operator-on-lists-why-do-these-give-different-results).

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,142 @@
...
+    
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"] += [movie]
```
Comment (yangashley) [inline viewing_party/party.py:19]: See comment above about using `append`

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,142 @@
...
+    return user_data
 
+def watch_movie(user_data, title):
+    for data in user_data["watchlist"]:
+        if data["title"] == title:
+            user_data["watchlist"].remove(data)
+            user_data["watched"].append(data)
+            break
```
Comment (yangashley) [inline viewing_party/party.py:27]: 🙌 Nice job using a `break` in the loop here

Here's a [resource](https://support.techsmart.codes/hc/en-us/articles/360058068873-For-Loops-Why-shouldn-t-I-add-to-or-remove-from-a-list-during-a-for-each-loop) about why we shouldn't mutate a collection while iterating over it if you'd like to read more.

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,142 @@
...
+def add_to_watched(user_data, movie):
+    user_data["watched"] += [movie]
+    return user_data
+    
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"] += [movie]
+    return user_data
 
+def watch_movie(user_data, title):
+    for data in user_data["watchlist"]:
```
Comment (yangashley) [inline viewing_party/party.py:23]: Consider naming the looping variable `data` to something more descriptive. What is the element in the watchlist? Since it's a movie, we could rename the variable `movie`. 

```python
for movie in user_data["watchlist"]:

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,142 @@
...
+            break
+    return user_data
+    
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum_ratings = 0 
+    if len(user_data["watched"]) == 0:
+        return 0
```
Comment (yangashley) [inline viewing_party/party.py:36]: If a the length of the watched list is 0, then `sum_ratings` gets created and then nothing happens to it. We should instantiate the variable when we anticipate it being used, which is after your if-statement. 

For line 35, it would be more Pythonic to write `if not user_data["watched"]`.

Also, per the requirements in the README, the function should return 0.0 if the watched list is empty.

```python
def get_watched_avg_rating(user_data):
    if not user_data["watched"]:
        return 0.0

    sum_ratings = 0
    for watched_movie in user_data["watched"]: 
        sum_ratings += watched_movie["rating"]
    return sum_ratings / len(user_data["watched"])

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,142 @@
...
+        sum_ratings += watched_movie["rating"]
+    return sum_ratings / len(user_data["watched"])
+
+def get_most_watched_genre(user_data):
+    genre_dict = {}
+    for movie in user_data["watched"]:
+        if movie["genre"] in genre_dict:
+            genre_dict[movie["genre"]] += 1
+        else:
+            genre_dict[movie["genre"]] = 1
```
Comment (yangashley) [inline viewing_party/party.py:47]: You'll commonly see this pattern written like this when creating a frequency table:
```python
genre_dict[movie["genre"]] = movie.get("genre", 0) + 1

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,142 @@
...
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    user_watched_list = user_data["watched"].copy()
+    for friend in user_data["friends"]:
+        for friend_watched_movie in friend["watched"]:
+            if friend_watched_movie in user_watched_list:
+                user_watched_list.remove(friend_watched_movie)
```
Comment (yangashley) [inline viewing_party/party.py:69]: Instead of needing to make a copy of the entire `user_watched_list` and then removing movies from it, what if you created an empty list instead and added the unique movies to it?

Recall that the time complexity of `remove` is linear but the it's constant for `append`

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,142 @@
...
+    return user_watched_list
+
+def get_friends_unique_watched(user_data):
+    movies_watched_by_friend_only_list = []
+    for friend in user_data["friends"]:
+        for friend_watched_movie in friend["watched"]:
```
Comment (yangashley) [inline viewing_party/party.py:75]: Notice that lines 74-74 repeat what's written on lines 66-67. Consider creating a helper function that returns a list of the movies that friends have watched (maybe something like `get_friends_watched_movies`). Then you can invoke that helper function in `get_unique_watched` and here in `get_friends_unique_watched`. 

Not only does this DRY up your code, but it also makes these methods single responsibility. Instead of having to get the friends' movies and then decide if the movie is unique in someway, the logic can just be kept to determining uniqueness.

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,142 @@
...
+            if friend_watched_movie in movies_watched_by_friend_only_list:
+                continue
+            if friend_watched_movie not in user_data["watched"]:
```
Comment (yangashley) [inline viewing_party/party.py:78]: Nice job handling the scenario where a movie has been watched by more than one friend. How could you refactor this logic so that you don't have two if statements where one continues?

Could you remove lines 76-77 and update the if-statement on line 78 to check another condition as well?

Maybe something like:
```python 
if friend_watched_movie not in user_data["watched"] and friend_watched_movie not in movies_watched_by_friend_only_list:

```

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,142 @@
...
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    recommendations = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
```
Comment (yangashley) [inline viewing_party/party.py:93]: Yet another place where we have to get the movies that friends have watched! We could call that hypothetical helper function `get_friends_watched_movies` here

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,142 @@
...
+                continue
+            
+            elif movie in user_data["watched"]:
+                continue
+            
+            elif movie["host"] not in user_data["subscriptions"]:
+                continue
+            
+            else: 
+                recommendations.append(movie)
```
Comment (yangashley) [inline viewing_party/party.py:104]: Needing to have 3 conditions that use `continue` doesn't really convey the intent of what is going on here to me and impacts this function's readability. A good test of if something is readable is, to ask yourself if you think you could succinctly explain what's happening in a function to someone in a week. For me, I had to use the debugger to understand the logic. Ideally, we'd be able to understand what's happening by just reading. 

Consider reusing the function you already wrote called `get_friends_unique_watched` to get a list of movies only friends have seen. Then you can check for each movie if the value for "host" is in the user's subscriptions list, if so then add the movie to `recommendations`. What would that code look like?

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,142 @@
...
+
+    return recommendations
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    
+    favorite_genre = get_most_watched_genre(user_data)
```
Comment (yangashley) [inline viewing_party/party.py:114]: 👍 great job using a function you already wrote to keep your code DRY.

File: viewing_party/party.py [hunk 21]
```diff
@@ -1,23 +1,142 @@
...
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["genre"] != favorite_genre: 
+                continue
+            
+            elif movie in recommended_movies:
+                continue
+
+            elif movie in user_data["watched"]:
+                continue
```
Comment (yangashley) [inline viewing_party/party.py:126]: Like my comment above, the three `continue`s give me pause.

You invoked the method to get the most watched genre on line 114. You can also invoke the method `get_friends_unique_watched` to make the code more concise. 

When you have those two values, you can check if each movie in the friends_unique_movies list has a genre equal to the value of the most watched genre. If they're equal, then you can append that movie to recommended_movies.

Something like 
```python
recommended_movies = []
favorite_genre = get_most_watched_genre(user_data)
friends_unique_movies = get_friends_unique_watched(user_data)

for movie in friends_unique_movies:
        if favorite_genre == movie["genre"]:
            recommended_movies.append(movie)
return recommended_movies

File: viewing_party/party.py [hunk 22]
```diff
@@ -1,23 +1,142 @@
...
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    recommended_list = []
+    new_list = user_data["favorites"].copy()
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            if friend_movie in new_list:
+                new_list.remove(friend_movie)
```
Comment (yangashley) [inline viewing_party/party.py:139]: Since `remove` has linear time complexity, how can we go about writing the logic so that we append a movie to the list instead?

If you get a list of the movies watched by friends (with a helper method like suggested above) and then check that a user's favorite movies are not in that list, then we could append that movie to a new list like `recommendations`. 

Then you wouldn't need a whole copy of `user_data["favorites"]` and would not incur the linear time complexity of removing an element from a list (you'll have the constant time complexity of `append`).

This might look like:
```python
def get_rec_from_favorites():
    recommendations = []
    friends_watched = get_friends_watched_movies()

    for favorite in user_data["favorites"]:
        if not favorite in friends_watched:
           recommendations.append(favorite)
    return recommendations

File: viewing_party/party.py [hunk 23]
```diff
@@ -1,23 +1,142 @@
...
+
+            elif movie in user_data["watched"]:
+                continue
+
+            else: 
+                recommended_movies.append(movie)
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    recommended_list = []
```
Comment (yangashley) [inline viewing_party/party.py:133]: `recommended_list` is created but never used. Delete objects that are not accessed.

File: viewing_party/party.py [hunk 24]
```diff
@@ -1,23 +1,142 @@
+
```
Comment (yangashley) [inline viewing_party/party.py:1]: We should take care to not add in unnecessary changes to a commit so that our project stays neat. Looks like an unneeded blank line was added on line 1 here.

**Review by yangashley — COMMENTED** [review]
Nice work on viewing-party, Brianna and Vlada!

I also see that you two made frequent commits, nice job! In the future, as you continue to make small, frequent commits, I'd encourage you to practice writing concise and descriptive messages so when someone reviews the commit history there is a clear story of what happened over time.

Let me know if you have any questions about my review comments.
